heartbeat/security/security.go

```diff
@@ -46,6 +47,9 @@ func init() {
 	// The beat should use `getcap` at a later point to examine available capabilities
 	// rather than relying on errors from `setcap`
 	_ = setCapabilities()
+
+	// Make heartbeat dumpable so elastic-agent can access process metrics.
```

Comment:
Question: Is this applicable only when 
- We are not running as root, common in context of container (BEAT_SETUID_AS) and under elastic-agent. 
- We effectively switched the user id and group id which changed the dumpable flag to value other than 1 (restircting access). 

With that said, Should we enforce this change only on containerized environment? I assume there is no harm in overriding the dumpable flag even if its already 1?

Reply:
> I assume there is no harm in overriding the dumpable flag even if its already 1?

There shouldn't be, there's no issue with overriding and in case of error, kernel returns an error code we can safely ignore since the agent is ready to handle those cases [now](https://github.com/elastic/elastic-agent-system-metrics/pull/129).

> We are not running as root, common in context of container (BEAT_SETUID_AS) and under elastic-agent.
We effectively switched the user id and group id which changed the dumpable flag to value other than 1 (restircting access).

Not exactly, elastic-agent image is already shipped with file capabilities enabled so it doesn't require `BEAT_SETUID_AS` mechanism to kick in, see this point in prctl:

```
 •  The process executes ([execve(2)](https://man7.org/linux/man-pages/man2/execve.2.html)) a program that has
    file capabilities (see [capabilities(7)](https://man7.org/linux/man-pages/man7/capabilities.7.html)), but only if
    the permitted capabilities gained exceed those already
    permitted for the process.
```

Although changing effective uid/gid _also_ triggers disabling dumpable flag.

> Should we enforce this change only on containerized environment?

Container detection can be [unreliable](https://github.com/elastic/beats/pull/37794) in some cases. Also, I can't think of a situation where this would broaden the permission scope for a non-containerized instance (running users already have access to config files). But I'm happy to discuss alternatives.

Reply:
Gotcha, Thanks for the extra context. I agree with the argument here on container detection and permission scope. I think the workaround looks good as it is.